# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [2]:
import numpy as np
import pandas as pd


# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [3]:
from sklearn.datasets import load_diabetes
diabets = load_diabetes() 
diabets


{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [4]:
diabets.keys()


dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [5]:
print(diabets['DESCR'])


.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

1- First 10 columns are numeric predictive values. We label them as features.

2- diabetes['data'] is the X dataset, stores only feature values and diabetes['target'] is the y dataset and stores the lables for each date point.

3- 442

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [6]:
diabets['data'].shape # 443 data points

(442, 10)

In [7]:
diabets['target'].shape # 443 labels for each data points

(442,)

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [8]:
from sklearn.linear_model import  LinearRegression


#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [9]:
diabets_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [10]:
diabets_data_train = diabets['data'][:-20]
diabets_target_train = diabets['target'][:-20]
diabets_data_test = diabets['data'][-20:]
diabets_target_test = diabets['target'][-20:]


Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [11]:
diabets_model = diabets_model.fit(diabets_data_train, diabets_target_train)
print(f"coefficients equal to:  \n {diabets_model.coef_}.")
print(f"intercept equals to: {diabets_model.intercept_}.")


coefficients equal to:  
 [ 3.06094248e-01 -2.37635570e+02  5.10538048e+02  3.27729878e+02
 -8.14111926e+02  4.92799595e+02  1.02841240e+02  1.84603496e+02
  7.43509388e+02  7.60966464e+01].
intercept equals to: 152.76429169049118.


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [12]:
prediction = diabets_model.predict(diabets_data_test)


#### Print your `diabetes_target_test` and compare with the prediction. 

In [13]:
print('Actual labels are:')
print(diabets_target_test)

print('predicted labels are:')
print(prediction)

Actual labels are:
[233.  91. 111. 152. 120.  67. 310.  94. 183.  66. 173.  72.  49.  64.
  48. 178. 104. 132. 220.  57.]
predicted labels are:
[197.61898486 155.44031962 172.88875144 111.53270645 164.79397301
 131.06765869 259.12441219 100.47873746 117.06005372 124.30261597
 218.36868146  61.19581944 132.24837933 120.33293546  52.54513009
 194.03746764 102.5756431  123.56778709 211.03465323  52.60221696]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

No, the values in diabetes_target_test don't match the predicted values, and the reason is that the predicted values, are the output of a linear model, that tries to fit the training data. So there are two assumptions here, first the train data follows a linear pattern and secondly we could find the exact pattern with a linear regression model, and we now that none of these are 100 percent accurate so we have errors.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [14]:
# Relative path doesn't work for me
auto = pd.read_csv(r'D:\IronHack\labs\AI\lab-supervised-learning-regression\auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [15]:
auto.sample(5)

mpg  cylinders  displacement  horse_power  weight  acceleration  \
104  12.0          8         400.0        167.0    4906          12.5   
281  19.8          6         200.0         85.0    2990          18.2   
321  32.2          4         108.0         75.0    2265          15.2   
343  39.1          4          79.0         58.0    1755          16.9   
237  30.5          4          98.0         63.0    2051          17.0   

     model_year                car_name  
104          73        \t"ford country"  
281          79    \t"mercury zephyr 6"  
321          80      \t"toyota corolla"  
343          81      \t"toyota starlet"  
237          77  \t"chevrolet chevette"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [16]:
auto.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horse_power   392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   car_name      398 non-null    object 
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


What is the newest model year and the oldest model year?

In [17]:
print(f"The oldest car was made in the year of {auto['model_year'].min()}")
print(f"The newest car was made in the year of {auto['model_year'].max()}")

The oldest car was made in the year of 70
The newest car was made in the year of 82


Check the dataset for missing values and remove all rows containing at least one missing value.

In [18]:
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [19]:
auto.dropna(axis=0, how='any', inplace=True)

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [20]:
auto['cylinders'].value_counts()

cylinders
4    199
8    103
6     83
3      4
5      3
Name: count, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [21]:
auto.drop(['car_name'], axis=1, inplace=True)

In [22]:
from sklearn.model_selection import train_test_split
features = auto.drop(['mpg'], axis=1)
labels = auto['mpg']

X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = 0.2, random_state=42)


In [23]:
features

cylinders  displacement  horse_power  weight  acceleration  model_year
0            8         307.0        130.0    3504          12.0          70
1            8         350.0        165.0    3693          11.5          70
2            8         318.0        150.0    3436          11.0          70
3            8         304.0        150.0    3433          12.0          70
4            8         302.0        140.0    3449          10.5          70
..         ...           ...          ...     ...           ...         ...
393          4         140.0         86.0    2790          15.6          82
394          4          97.0         52.0    2130          24.6          82
395          4         135.0         84.0    2295          11.6          82
396          4         120.0         79.0    2625          18.6          82
397          4         119.0         82.0    2720          19.4          82

[392 rows x 6 columns]

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [24]:
model_auto_regr = LinearRegression()
model_auto_regr = model_auto_regr.fit(X_train, y_train)

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). In the end, we want the r-squared score to be as high as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [25]:
from sklearn.metrics import r2_score
y_pred = model_auto_regr.predict(X_train)
r2_score(y_train, y_pred)

0.8107227953093896

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [26]:
y_test_pred = model_auto_regr.predict(X_test)
r2_score(y_test, y_test_pred)

0.7942349075428587

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [43]:
X_train09, X_test09, y_train09, y_test09 = train_test_split(features, labels, test_size=0.1, random_state=1)


Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [44]:
auto_model09 = LinearRegression()
auto_model09 = auto_model09.fit(X_train09, y_train09)


Compute the predicted values and r squared score for our new model and new sample data.

In [45]:
y_pred09 = auto_model09.predict(X_test09)


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [46]:
r2_score(y_test09, y_pred09)


0.808502519021851

It improved slightly. I think considering the size of the data set(442), adding 10% of data points to the train dataset(42), doesn't really help the model. 

# Bonus Challenge - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [31]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [32]:
# Your code here:


Fit the model and print the ranking

In [33]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [34]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [35]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.